NDTensors/src/dense/dense.jl

```diff
@@ -3,70 +3,101 @@
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT())
+end
+
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(zeros(DataT, dim(inds)))
+end
+
+function Dense{ElT,DataT}(
+  ::UndefInitializer, inds::Tuple
+) where {ElT,DataT<:AbstractArray{ElT}}
+  return Dense{ElT,DataT}(DataT(undef, dim(inds)))
+end
+
+function Dense{DataT}() where {DataT<:AbstractArray}
+  return default_storagetype(DataT)()
+end
+
+function Dense{DataT}(inds::Tuple) where {DataT<:AbstractArray}
+  return default_storagetype(DataT)(inds)
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+function Dense{DataT}(::UndefInitializer, inds::Tuple) where {DataT<:AbstractArray}
+  return default_storagetype(DataT)(undef, inds)
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+function Dense{DataT}(x::Number, dim::Integer) where {DataT<:AbstractArray}
+  return Dense(fill!(DataT{typeof(x)}(undef, dim), x))
 end
 
+# This function is ill-defined. It cannot transform a complex type to real...
 function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
-  return ElT == ElR ? Dense(data) : Dense(ElR.(data))
+  return Dense{ElR,similartype(typeof(data), ElR)}(data)
 end
 
-# Construct from a set of indices
-function Dense{ElT,VecT}(inds) where {ElT,VecT<:AbstractArray{ElT}}
-  return Dense(VecT(dim(inds)))
+function Dense{ElT}(dim::Integer) where {ElT<:Number}
+  return default_storagetype(ElT)(zeros(default_datatype(ElT), dim))
 end
 
-Dense{ElT}(dim::Integer) where {ElT} = Dense(zeros(ElT, dim))
+Dense{ElT}() where {ElT} = default_storagetype(ElT)()
+
+function Dense(data::DataT) where {DataT<:AbstractArray{ElT,N}} where {ElT,N}
+  if N > 1
+    Dense(vec(data))
+  else
+    Dense{ElT,DataT}(data)
+  end
+end
 
-Dense{ElT}(::UndefInitializer, dim::Integer) where {ElT} = Dense(Vector{ElT}(undef, dim))
+function Dense(DataT::Type{<:AbstractArray{ElT}}, dim::Integer) where {ElT}
+  return Dense{ElT,DataT}((dim,))
+end
 
 Dense(::Type{ElT}, dim::Integer) where {ElT} = Dense{ElT}(dim)
 
-Dense(x::ElT, dim::Integer) where {ElT<:Number} = Dense(fill(x, dim))
-
-Dense(dim::Integer) = Dense(Float64, dim)
+function Dense(ElT::Type{<:Number}, ::UndefInitializer, dim::Integer)
+  return Dense{ElT,default_datatype(ElT)}(undef, (dim,))
+end
 
-Dense(::Type{ElT}, ::UndefInitializer, dim::Integer) where {ElT} = Dense{ElT}(undef, dim)
+Dense(x::Number, dim::Integer) = Dense(fill!(default_datatype(typeof(x))(undef, dim), x))
 
-Dense(::UndefInitializer, dim::Integer) = Dense(Float64, undef, dim)
+Dense(dim::Integer) = Dense(default_eltype(), dim)
 
-Dense{ElT}() where {ElT} = Dense(ElT[])
 Dense(::Type{ElT}) where {ElT} = Dense{ElT}()
```

Comment:
Should this be specialized to `Number`, i.e.:
```julia
Dense(eltype::Type{<:Number}) = Dense{eltype}()
```

Reply:
Still seems like we need to constrain this, otherwise it would be ambiguous with calls like `Dense(Vector{Float32})`.